### PR TITLE
Forts & Citadels

### PR DESCRIPTION
**Pull request by r3versi**
New tile sprite for Fort and Citadel by @GeneralWadaling

### CODE REVIEW

File: core/src/com/unciv/logic/battle/BattleDamage.kt [hunk 1]
```diff
@@ -154,11 +155,18 @@ class BattleDamage{
...
+        modifiers.putAll(getTileSpecificModifiers(defender, tile))
 
         if (!defender.unit.hasUnique("No defensive terrain bonus")) {
-            val tileDefenceBonus = defender.getTile().getDefensiveBonus()
-            if (tileDefenceBonus > 0) modifiers["Terrain"] = tileDefenceBonus
+            val tileDefenceBonus = tile.getDefensiveBonus()
+            if (tileDefenceBonus > 0)
+                modifiers["Terrain"] = tileDefenceBonus
+
+            val isFriendlyTerritory = tile.getOwner() != null && !defender.getCivInfo().isAtWarWith(tile.getOwner()!!)
```
Comment (yairm210) [inline core/src/com/unciv/logic/battle/BattleDamage.kt]: Pretty sure there's a function like this when checking for healing as well

File: core/src/com/unciv/logic/map/MapUnit.kt [hunk 2]
```diff
@@ -631,5 +633,21 @@ class MapUnit {
...
+    private fun getCitadelDamage() {
+        // Check for Citadel damage
+        val applyCitadelDamage = currentTile.neighbors
+                .filter{ it.getOwner() != null && civInfo.isAtWarWith(it.getOwner()!!) }
+                .map{ it.getTileImprovement() }
+                .filter{ it != null && it.hasUnique("Deal 3 damage to adjacent enemy units") }
+                .any()
+
+        if (applyCitadelDamage) {
+            health -= 3
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/MapUnit.kt]: Probably 30, no?

File: core/src/com/unciv/logic/battle/BattleDamage.kt [hunk 3]
```diff
@@ -154,11 +155,18 @@ class BattleDamage{
...
+            val improvement = tile.getTileImprovement()
+            if (isFriendlyTerritory && improvement != null)
+                if (improvement.hasUnique("Gives a defensive bonus of 50%")) modifiers["Fort"] = 0.50f
```
Comment (yairm210) [inline core/src/com/unciv/logic/battle/BattleDamage.kt]: Modifiers[improvement.name]=

File: core/src/com/unciv/logic/map/MapUnit.kt (hunk 2 above)
Comment (r3versi) [reply core/src/com/unciv/logic/map/MapUnit.kt]: 3 in vanilla, 30 in G&K [source](https://civilization.fandom.com/wiki/Citadel_(Civ5))

Comment (yairm210) [reply core/src/com/unciv/logic/map/MapUnit.kt]: Unciv works with a 100-health point system, so that'll be 30 =)

Comment (r3versi) [reply core/src/com/unciv/logic/map/MapUnit.kt]: What you think we should do? Keep this to 3 and wait for G&K to update it to 30? 3 seems a really low value, 30 seems a really big one though :thinking:

**Comment by r3versi** [comment]
I added a method `TileInfo.isFriendlyTerritory(civInfo: CivInfo)` that performs the same checks that were performed in `rankTileForHealing`.

> Add you've probably seen, there's a pr that's been waiting for months regarding forts.
> The reason it wasn't merged is primarily because this is to big of a human cheat.
> The AI needs to know how to handle forts and citadels, at the very least, and for bonus points where and when to place them.

Now I remember why I had this PR on standby for weeks.. :laughing:
I'll keep you updated if I can come up with an idea.

**Comment by JackRainy** [comment]
I would suggest the next algorithm:
Forts
The range of 2 tiles is chosen to be able to attack it by city, in case the enemies have captured the fort, and far enough to be the last line of defence, not allowing catapults and artillery reach closer.

**Comment by JackRainy** [comment]
A note to be mentioned: "Forts can be constructed on Bananas without removing the jungle." Did we take that into account?
https://civilization.fandom.com/wiki/Bananas_(Civ5)

**Comment by RobertGine** [comment]
Correct me if I'm wrong but by Jack's algorithm, the AI will build as many Forts as possible next to the closest city to the enemy, but will not defend other cities (that might also be vulnerable)?

**Comment by JackRainy** [comment]
@RobertGine, it depends what you mean by "the closest city to the enemy".
If the enemy approaches to a city, it builds the defence, if the enemy goes to another city, they build the defence there.

**Comment by RobertGine** [comment]
Oh. So the AI will throw workers in front of an attacking army?

**Comment by JackRainy** [comment]
I understand what you mean :)
That's why they should start building it a bit earlier - while relations only become bad. Same time when AI decides to build army.

**Comment by yairm210** [comment]
#2325 adds this, with AI =)

**Pull request closed without merging**